### ISSUE TITLE
[React] CSS not loading when using redirectTo

### ISSUE THREAD

**Issue by mparker11** [pshrmn/curi#125]
**Version - @curi/router: 1.0.4**

_Disclaimer, I'm using SSR so I'm not sure if that's related._

I'm trying to use `redirectTo` in my routes file by doing this
```
{
        name: 'CreateAccount',
        path: 'signup',
        resolve: {
            authorized: hasPlanSelected //this is a separate function returning a promise
        },
        response({ resolved }) {
            if (resolved.authorized) {
                return { body: CreateAccount };
            }

            return {
                redirectTo: { name: 'Pricing' },
                status: 302
            };
        }
    },
```

It's working, but once I'm redirected to `Pricing`, the CSS is not loaded although I'm importing it at the top of the file. I took a look at `@hickory/browser` and found `finalizeReplace` (line 65), and I see that there's a try/catch block using `window.history.replaceState` first. Would this have anything to do with CSS not loading for a new page? 

It's my hunch because when I manually use `window.location.replace` as the redirect tool, the CSS is normal.

**Comment by pshrmn** [comment]
`window.location.replace` causes a full page load (hits the server). It is there for a fallback if `replaceState()` fails for some reason, but should almost never be necessary.

Is your application doing dynamic CSS loading? Can you open the network tab (in chrome's dev tools) and see if there is a request for the expected file?

**Comment by mparker11** [comment]
I'm using Create React App (no eject) so all pages and CSS are bundled in one JS file. No specific CSS files are requested in the Network tab. Maybe this is outside of your project. It seems like a thing with React and replacing history state, but I'm not too sure. The only reason it makes sense to me is because I have never used `replaceState` when redirecting.

**Comment by pshrmn** [comment]
(It may not be directly related, but it always helps to know what issues people run into while using Curi.)

Can you expand on what you mean by "the CSS is not loaded"? If I understand CRA correctly, it injects the styles into the page's `<head>`. Is that not happening in your application? Also, when you say SSR, are you calling `yarn build` and running a local server or using `yarn start`?

It just seems odd to me that that isn't working. Typically when there are issues with static content in SPAs, it is because a relative path is used instead of an absolute path (so a request is made for an invalid file), but that shouldn't be the case here.

There are a few situations that I am curious about:

1. When the user is authorized:
  a. Navigating directly to `CreateAccount`
  b. Navigating to `CreateAccount` from another page in your app
2. When the user is **not** authorized
  a. Navigating directly to `CreateAccount`
  b. Navigating to `CreateAccount` from another page in your app
3. Navigating directly to `Pricing`

Based on what you described above, I believe that 2a does not work, but I am unsure about the other ones.

**Comment by mparker11** [comment]
One thing I did differently is use `emitRedirects: false` on the router's init. Would that have anything to do with it? I was confused by the documentation on what it's used for. I just know without it, on redirect I get the React error `type is invalid -- expected a string or a class/function but got: undefined`.

**Comment by pshrmn** [comment]
The app is setup with CRA, but you're running your own server? Something like this [example](https://github.com/pshrmn/curi/blob/3b70a9d44a9fead80e27984203872169daa66b61/examples/misc/server-rendering/renderer.js)? Does the HTML (e.g. view source) look like what you're expecting?

To be honest, I'm a bit stumped based on what you described. I don't understand how the application is running (so that Curi runs and it knows to redirect) but the bundle isn't inserting the styles into the `<head>`.

I _think_ that if your server sends a `302` response, the browser should automatically redirect if you get that setup correctly. That would get around the issue that you are having, although I would still be curious to understand why the CSS isn't working.

I'll expand a bit on redirects so that hopefully you can verify everything is setup as expected on the server.

---

The idea with `emitRedirects` is that if a response only has a `redirectTo` property, your render function probably doesn't know how to render it (since there is no `response.body` component). When `emitRedirects` is `false`, the render function won't be called for that response, so you don't have to worry about trying to render something for it.

```jsx
function render({ response }) {
  // this only works if response.body exists
  const { body:Body } = response;
  return <Body response={response} />
}
```

You _could_ return a response that has a `body` property, but when you're going to re-render for the redirected location right away, it isn't really necessary.

`emitRedirects` should be `true` (the default) on the server, since you would want to detect that and set the status on the response (and you don't even have to render the application).

For server-side rendering, you also want to set `automaticRedirects` to `false`. That was the result a more recent discovery where the response for the redirect was generated before `router.once()` is called.

```js
function(req, res) {
  const history = InMemory({ locations: [req.path] });
  const router = curi(history, routes, {
    automaticRedirects: false
  });
  router.once(({ response }) => {
    if (response.status === 302) {
      return res
        .redirect(302, response.redirectTo.pathname)
        .send();
    }
    // ...
  });
}
```

---

I should probably put together a guide on redirecting and a guide on server side rendering. I think that Curi handles both of those well, but they are also little bit tricky to reason about, so the docs would benefit from adding that information.

**Comment by mparker11** [comment]
>The app is setup with CRA, but you're running your own server? Something like this example?

Yep, very identical to that example. I started doing that months ago when I first started learning about server side rendering because CRA wasn't quite up to par. It's worked well. 

>Does the HTML (e.g. view source) look like what you're expecting?

Yep. That's all good. Every bit of CSS is in the `<head>` and the views switch as needed.

>I think that if your server sends a 302 response, the browser should automatically redirect if you get that setup correctly.

This is where I'm lacking knowledge on routing. Since I'm basing my authorization for a page on the client, how can I get the server to redirect? I'm guessing if I can make that happen, then the CSS will have no trouble rendering. But at this point, the server renders `CreateAccount`, and never `Pricing` when redirecting. As I'm typing, it makes sense conceptually, but I don't understand what the solution would be because I haven't developed on the server enough.

**Comment by mparker11** [comment]
Also, if it helps, my setup includes [React App Tools](https://www.npmjs.com/package/react-app-tools) which is how I'm running the server.

**Comment by pshrmn** [comment]
I am curious what happens if you configure the server to not do SSR; will navigating directly to `CreateAccount` work on the client side? To test that, I would either save an `index.html` file and use `res.sendFile()` to respond with that or use whatever inline HTML generation you are doing without inserting the React generated HTML.

```js
function render(req, res) {
  res.sendFile("./public/index.html");
}

// re-using your markup insert function might be faster

function render(req, res) {
  res.send(markup(""));
}

function markup(html) {
  return `<!doctype html>
<html>
<body>
  <div id="app">${html}</div>
  <script src="/static/bundle.js"></script>
</body>
</html>`;
}
```
If it works when Curi is only running on the client, that should confirm that the issue originates with the server.

**Comment by mparker11** [comment]
Your suggestion isn't quite working for me because of how [React App Tools](https://www.npmjs.com/package/react-app-tools) works, but I worked around it with a change to `package.json` and got it to work client-side only.

The problem definitely lies on the server rendering. The redirect works as expected using Curi. Any other ideas on a solution for the server? I know it would be difficult since I'm using React App Tools and you didn't build it but any more insight would be helpful. Thank you for all your help!

**Comment by pshrmn** [comment]
Have you tried updating the handler to send a `302 redirect` when the response has a `redirectTo` property?
```js
function handler(req, res) {
  // ...
  const router = curi(history, routes, {
    automaticRedirects: false
  });
  router.once(({ response }) => {
    if (response.redirectTo) {
      res.redirect(302, response.redirectTo.pathname);
      return;
    }
    const html = renderToString(...)
    // ...
  });
}
```
or maybe your `CreateAccount` route is running into issues while creating a response? Maybe check the `error` property in the `response()` function or add a log statement in `router.once()`.

```js
{
  name: 'CreateAccount',
  path: 'signup',
  resolve: {
    authorized: hasPlanSelected //this is a separate function returning a promise
  },
  response({ error, resolved }) {
    if (error) {
      // ...
    }      
    if (resolved.authorized) {
      return { body: CreateAccount };
    }
    return {
      redirectTo: { name: 'Pricing' },
      status: 302
    };
  }
},
```
```js
function renderHandler(req, res) {
  // ...
  router.once(({ response }) => {
    console.log({ path: req.path, response }); // see if this is what you're expecting
  });
}
```

**Comment by mparker11** [comment]
There is no error from the route on Curi but it's odd, if I add `console.log({ path: req.path })` to my `router.once` call, I only get `{ path: '/signup' }` once, and nothing shows when it redirects. So my question is how do we get Curi to communicate to the server that a redirect is happening? Here is my code for the server for context:
```
const path = require('path');
const express = require('express');
const React = require('react');
const { renderToString } = require('react-dom/server');

const { curi } = require('@curi/router');
const InMemory = require('@hickory/in-memory');
const { curiProvider } = require('@curi/react-dom');
const { default: routes } = require('../routes');
const { default: AppWrapper } = require('../containers/AppWrapper');
const { default: redirects } = require('../redirects');
const assets = require('./assets.json');

const app = express();
app.use(express.static(path.join(__dirname, '/public')));

app.get('*', (req, res) => {
    const history = InMemory({ locations: [req.url] });
    const router = curi(history, routes, {
        automaticRedirects: false
    });
    const Router = curiProvider(router);

    router.once(({ response }) => {
        console.log({ path: req.path });
        if (response.status === 302) {
            return res
                .redirect(302, response.redirectTo.pathname)
                .send();
        }

        const pageMarkup = renderToString(
            <Router>
                {() => (<AppWrapper />)}
            </Router>
        );

        res.send(`
            <html>
                <head>
                    <!-- The first thing in any HTML file should be the charset -->
                    <meta charset="utf-8">
                    <!-- Make the page mobile compatible -->
                    <meta name="viewport" content="width=device-width, initial-scale=1">
                    <!-- Allow installing the app to the homescreen -->
                    <link rel="manifest" href="manifest.json">
                </head>
                <body>
                    <div id="root">${pageMarkup}</div>
                    ${
                        Object.keys(assets).map((key) => {
                            let mapped = assets[key].map((src) => {
                                return `<script src="/${src}"></script>`
                            });
                            return mapped[0];
                        }).join('')
                    }
                </body>
            </html>
        `);
    });
});

if (process.env.NODE_ENV === 'production') {
    app.listen(process.env.PORT || 8080);
} else {
    module.exports.default = app;
}

```

**Comment by pshrmn** [comment]
Did you inspect the `response`? It turns out `res.redirect().send()` isn't a thing (you don't need the `send()`), so if you aren't getting an error, I assume that the `response` doesn't have a `redirectTo` property (i.e. `resolved.authorized` must be `true`).

**Comment by mparker11** [comment]
Yeah the response is not being sent back on redirect. Notice that I have a `console.log` right before `if (response.status === 302)`. It only prints on first load, but never the redirect.

**Comment by mparker11** [comment]
I figured it out! I'm not sure if it's ideal but basically I told the client side router to do the redirect.
```
router.once(({ response }) => {
    if (response.status === 302) {
        window.location.replace(response.redirectTo.pathname);
        return;
    }
    ReactDOM.hydrate((
        <Router>
            {() => ( <AppWrapper /> )}
        </Router>
    ), root);
});
```

This way the server is hit again. But if there's something better, please feel free to let me know. Thanks again for all your help!

**Comment by pshrmn** [comment]
I feel like I'm missing something because you should not have to call `window.location.replace()`. Unless you use `automaticRedirects: false` on the client, Curi should use `history.replaceState()` to change locations automatically.

That element type error happens because `response.body` is `undefined`. If you want to emit redirects, you can write a function component and attach it to the redirect response as its `body`.

```jsx
const Redirecting = () => <div>Redirecting...</div>;

response({ resolved }) {
  // ...
  return {
    redirectTo: { name: "Pricing" },
    status: 302,
    body: Redirecting
  };
}
```

As for the server, is there any situation where the user would navigate directly to `/signup` (so that the server handles a request for `/signup`) and they should not be redirected (i.e. calling `resolve.authorized()` on the server should return `true`)? If not, then that `res.redirect()` call should happen and the browser should redirect for you. When you call `window.location.replace()`, that is doing pretty much the same thing, but manually.

I updated the [server rendering example](https://github.com/pshrmn/curi/blob/0e6209ab2103fc03d9121fdbf76cde62cacb2637/examples/misc/server-rendering/renderer.js) with a redirect to verify that this approach works (admittedly for a much simpler use case).

---

As a side not, I'm realizing that `response.redirectTo` should have a `url` property. That way, if you include a query/hash in the redirect, you don't have to manually concatenate those strings together.

**Comment by mparker11** [comment]
>As for the server, is there any situation where the user would navigate directly to /signup (so that the server handles a request for /signup) and they should not be redirected

Nope. The first part of the flow is to choose a plan. Users will not be able to sign up without a plan selected.

>If you want to emit redirects, you can write a function component and attach it to the redirect response as its body.

It would be nice to understand how to make this work without emitting redirects but the problem remains that the server doesn't get the redirect call. Only the client side does. The elephant in the room is that I'm using React App Tools so that could be a blocker. Also, I'm trying to do research on how `pushState` affects the server, if at all, to get more clues.

**Comment by pshrmn** [comment]
`pushState`/`replaceState` navigation never hits the server (except for any API calls that may result from the navigation).

For reference, when my example server sends a redirect, this is what my network tab looks like:

![ssr-redirect](https://user-images.githubusercontent.com/1127037/47257565-26057280-d455-11e8-818e-124cf997dd1b.png)

I'm still a bit hung up on these lines from your server:

```js
if (response.status === 302) {
  return res
    .redirect(302, response.redirectTo.pathname)
    .send();
}
```
Have you edited that to get rid of the `.send()` call that throws an error? And do requests for `/signup` hit that loop? A super quick way to check if correctly redirecting will fix issues would be to hard code a `/signup` check at the top of the handler function.

```js
function renderHandler(req, res) {
  if (req.path === "/signup") {
    res.redirect(302, "/choose-plan"); // I'm not sure what the actual pathname is
    return;
}
```

**Comment by mparker11** [comment]
>Have you edited that to get rid of the .send() call that throws an error? And do requests for /signup hit that loop?

Yes, I removed that `.send()` call, but the error never happened because there is no second call to the server router to check for the 302 status. So my problem isn't redirecting from the server, it's when the redirect happen in Curi client side
```
{
        name: 'CreateAccount',
        path: 'signup',
        resolve: {
            authorized: hasPlanSelected
        },
        response({ resolved }) {
            if (resolved.authorized) {
                return { body: CreateAccount };
            }

            return {
                redirectTo: { name: 'Pricing' },
                status: 302
            };
        }
    },
```

and never hits the server check
```
app.get('*', (req, res) => {
    const history = InMemory({ locations: [req.url] });
    const router = curi(history, routes, {
        automaticRedirects: false
    });
    const Router = curiProvider(router);

    router.once(({ response }) => {
        if (response.redirectTo !== undefined) {
            res.redirect(302, response.redirectTo.pathname);
            return;
        }

        //everything else...
        //res.send()
    });
```

**Comment by mparker11** [comment]
Re-reading your comment...
>pushState/replaceState navigation never hits the server (except for any API calls that may result from the navigation).

So how are we suppose to catch a redirect on the server if `replaceState` never hits it?

**Comment by pshrmn** [comment]
With single-page applications, in-app navigation should never hit the server. The only request that the server handles is the initial page.

Direct navigation to `/signup`

1. User navigates directly to `/signup`.
2. The browser sends a request for `/signup` to the server.
3. Curi matches the `CreateAccount` route and calls `hasPlanSelected`.
4. `hasPlanSelected` returns `false` (right?) and returns a response with `{ status: 302, redirectTo: {...}}`.
5. The `router.once()` callback sees `response.redirectTo` and calls `res.redirect()`.
6. The browser receives the redirect and sends another request to the server for the redirect location.

In-app navigation to `/signup`

1. User navigates directly to `/`.
2. The browser sends a request for `/` to the server.
3. Curi matches the `Home` route and returns a response with `{ body: Home }`
4. The `router.once()` callback renders the application into an html string, which is inserted into the page HTML and sent to the user.
5. The user clicks a link to `/signup` and Curi intercepts the navigation.
6. Curi matches the navigation to its `CreateAccount` route and calls `hasPlanSelected`.
7. Based on the return value of `hasPlanSelected`:
  a. if `true`, the response will have a body and render the `/signup` page conteont.
  b. if `false`, the response will have a `redirectTo` property. Curi will see that and automatically redirect (using `history.replaceState`) to the new location. If `emitRedirects` is `true`, the response with the `redirectTo` property will be emitted (i.e. the React app will re-render, but you should only do this if the response also has a `body`). If `emitRedirets` is `false`, the response will not be emitted, so the app will not re-render until the response for the `Pricing` route is emitted.

(sorry, that got a bit wordy)

**Comment by pshrmn** [comment]
If your server calls `res.redirect()`, the redirect will happen automatically. In order for this to happen, the request for `/signup` has to result in a Curi response with a `redirectTo` property (in order to enter that `if` statement).

The `redirectTo` property only exists when `resolve.authorized` is `false` (`hasPlanSelected` returns `false`). I want to say that that function must be returning `true` on the server for some reason.

**Comment by mparker11** [comment]
I understand but the problem is that the server is never called again at all to enter the if statement. I'll try to demonstrate:
```javascript
app.get('*', (req, res) => {
    //init the router here...blah blah

    router.once(({ response }) => {
       //should log a response
        console.log(response)
        //^^^never logs a response on redirect before getting to the if statement^^^

        if (response.redirectTo !== undefined) {
            res.redirect(302, response.redirectTo.pathname);
            return;
        }

        //everything else...
        //res.send()
    });
```

**Comment by pshrmn** [comment]
> never logs a response on redirect

By this, do you mean when when `req.path` is `/signup`? Or is the browser receiving the `302 Redirect` and not automatically sending another response?

Also, did you try hard-coding the redirect at the top of the handler?

**Comment by mparker11** [comment]
>Also, did you try hard-coding the redirect at the top of the handler?

Yes, this works fine.

>By this, do you mean when when req.path is /signup? Or is the browser receiving the 302 Redirect and not automatically sending another response?

The browser is receiving the 302 redirect but is not automatically sending another response to the server. This is how I was able to detect and manually reload on the browser. But as you said, we want the server to know but nothing is being sent to it at all.

**Comment by pshrmn** [comment]
> The browser is receiving the 302 redirect but is not automatically sending another response to the server. This is how I was able to detect and manually reload on the browser.

You're seeing a `302 redirect` in the network tab (the server response) or the "response" that Curi creates has a `status: 302`?

**Comment by mparker11** [comment]
Nope. Here's a screenshot. My second route (`/pricing`) never shows.

![shot_181020_110924](https://user-images.githubusercontent.com/825507/47259034-b6918200-d458-11e8-9bfe-4b30a5c52b97.png)

**Comment by pshrmn** [comment]
When `/signup` returns a `200` (it should not), there will never be another request sent to the server. Curi will handle the redirect on the client side (using `history.replaceState`). However, `/signup` should not return a `200`.

I _really_ think the issue is because `resolve.authorized` is returning `true` on the server when it should always be returning `false`.

**Comment by mparker11** [comment]
>I really think the issue is because resolve.authorized is returning true on the server

Oh I see what you mean! Ok, you're right, it is true because on the server. I'm sorry for the confusion. Here's my method for `hasPlanSelected`:
```javascript
const hasPlanSelected = () => {
    if (process.browser) {
        if (!__isEmpty(localStorage.getItem('selectedPlan'))) {
            return Promise.resolve(true);
        } else {
            return Promise.resolve(false);
        }
    } else {
        //let SSR bypass
        return Promise.resolve(true);
    }
};
```

The reason I let the server come out as true is because `localStorage` is not ready on the server, so it caused an error. I guess the new question is, what is a good way for the server to know if there is a selected plan? I was trying to avoid any API calls for this simple prop.

**Comment by mparker11** [comment]
I've been researching and I had no idea that cookies were readable from the server. I think I found my answer. Yet again, you helped me solve an issue I caused. I really appreciate it.

**Comment by pshrmn** [comment]
Awesome! I was just writing a comment about using cookies because there isn't a great way to access them from `hasPlanSelected`. :smile:

The approach that I would take is to attach them to the location's `state`.

```js
function renderHandler(req, res) {
  const InMemory({ locations: [
    { pathname: req.path, state: req.cookies }
  ] });
}
```

`hasPlanSelected` will be passed the location, so you could check the cookies using that.

```js
const hasPlanSelected = ({ location }) => {
  if (process.browser) {
    // ...
  } else {
    // resolve true if the cookie exists
  }
};
```

As a style nit, `!__isEmpty(localStorage.getItem('selectedPlan'))` is a boolean, so you can change that `if/else` to:

```js
return Promise.resolve(
  !__isEmpty(localStorage.getItem('selectedPlan'))
)
```

**Comment by mparker11** [comment]
🙏🙏🙏🙏🙏
You have been amazing. I'll get this implemented. I thought I was done "figuring out" server-side rendering but I had a lot more to learn. Thanks so much again.

**Comment by pshrmn** [comment]
You're welcome! I'm working on putting together an SSR guide, so hopefully some of the setup will be more obvious.
